File: 第一章/learning_service/src/add_server.cpp
#include<ros/ros.h>
#include<learning_service/Add.h>

bool addCallback(learning_service::Add::Request  &req,
                 learning_service::Add::Response &res)
{
    res.sum = req.a + req.b;

    ROS_INFO("request: x=%0.2f, y=%0.2f", req.a, req.b);
    ROS_INFO("the result is: %0.2f", res.sum);

    return true;
}

int main(int argc, char **argv)
{
        ros::init(argc, argv, "add_server");

        ros::NodeHandle nh;

        ros::ServiceServer add_service = nh.advertiseService("add_service", addCallback);

        ROS_INFO("Ready to add two number.");

        ros::spin();

        return 0;
}

File: 第一章/learning_service/scripts/add_server.py
#!/usr/bin/env python
# coding:utf-8

import rospy
from learning_service.srv import Add, AddResponse

def addCallback(req):
    print("the request and resutle is: %0.2f + %0.2f = %0.2f"%(req.a, req.b, (req.a + req.b)))
    return AddResponse(req.a + req.b)

def add_server():
    rospy.init_node('add_server')
    s = rospy.Service('add_service', Add, addCallback)
    print("Ready to add two number.")
    rospy.spin()

if __name__ == "__main__":
    add_server()

